File: src/Sorting/QuickSort.java
package Sorting;

public class QuickSort {
    public static void main(String args[]){
        int[] array = {1,3,5,7,9,2,4,6,8,10};
        quickSort(array,0,array.length-1);
        for(int i=0;i<array.length;i++){
            System.out.println(array[i]);
        }
    }
    public static void quickSort(int[] arr,int si,int ei){
        if(si>=ei) return;

        int pivot = partition(arr,si,ei);
        quickSort(arr,si,pivot-1);
        quickSort(arr,pivot+1,ei);
    }

    private static int partition(int[] arr, int si, int ei) {
        int pivot = arr[si];
        int count = 0;
        for(int i = si+1;i<=ei;i++){
            if(arr[i]<pivot){
                count++;
            }
        }
        int pivot_pos = si+count;
        //swap(arr[si],arr[si+count]);
        int temp = arr[si];
        arr[si] = arr[si+count];
        arr[si+count] = temp;
        int i = si;
        int j = ei;
        while(i<pivot_pos && j>pivot_pos){
            if(arr[i]<arr[pivot_pos]){
                i++;
            }
            else{
                if(arr[j]<arr[pivot_pos]){
                    //swap(arr[i],arr[j]);
                    int t = arr[i];
                    arr[i] = arr[j];
                    arr[j] = t;
                }
                else{
                    j--;
                }
            }
        }
        return pivot_pos;
    }
    

}

File: Sorting/quicksort.py
def quickSort(arr, beg, end):
    # Please add your code here
    if beg >= end:
        return
    pivot = partition(arr, beg, end )
    quickSort(arr, beg, pivot-1)
    quickSort(arr, pivot+1, end)

def partition(arr, start, end):
    pivot = arr[start]
    count = 0
    for i in range(start+1, end+1):
        if arr[i] < pivot:
            count +=1
    pivot_pos = start + count
    arr[pivot_pos], arr[start] = arr[start], arr[pivot_pos]
    pivot = arr[pivot_pos]
    
    i = start
    j = end
    while i < pivot_pos and j > pivot_pos:
        if arr[i] <= pivot:
            i +=1
        else:
            if arr[j] >= pivot:
                j-=1
            else:
                arr[i], arr[j] = arr[j], arr[i]
    return pivot_pos

n=int(input())
arr=list(int(i) for i in input().strip().split(' '))
quickSort(arr, 0, n-1)
print(*arr)
